packages/headless/src/controllers/generated-answer/headless-generated-answer.ts

```diff
@@ -0,0 +1,139 @@
+  logRetryGeneratedAnswer,
+} from '../../features/generated-answer/generated-answer-analytics-actions';
+import {generatedAnswerReducer as generatedAnswer} from '../../features/generated-answer/generated-answer-slice';
+import {GeneratedAnswerState} from '../../features/generated-answer/generated-answer-state';
+import {executeSearch} from '../../features/search/search-actions';
+import {GeneratedAnswerSection} from '../../state/state-sections';
+import {loadReducerError} from '../../utils/errors';
+import {buildController} from '../controller/headless-controller';
+
+export type {GeneratedAnswerState, GeneratedAnswerCitation};
+
+/**
+ * @internal
+ */
+export interface GeneratedAnswer {
+  /**
+   * The state of the GeneratedAnswer controller.
+   */
+  state: GeneratedAnswerState;
+  /**
+   * Re-executes the last query to generate an answer.
+   */
+  retry(): void;
+  /**
+   * Determines if the generated answer was liked, or upvoted by the end user.
+   */
+  like(): void;
+  /**
+   * Determines if the generated answer was disliked, or downvoted by the end user.
+   */
+  dislike(): void;
+  /**
+   * Logs a custom event indicating a cited source link was clicked.
+   * @param id The ID of the clicked citation.
+   */
+  logCitationClick(id: string): void;
+}
+
+/**
+ * @internal
+ */
+export function buildGeneratedAnswer(engine: SearchEngine): GeneratedAnswer {
+  if (!loadGeneratedAnswerReducer(engine)) {
+    throw loadReducerError;
+  }
+
+  const {dispatch} = engine;
+  const controller = buildController(engine);
+  const getState = () => engine.state;
+
+  let abortController: AbortController | undefined;
+  let lastRequestId: string;
+  let lastStreamId: string;
+
+  const setAbortControllerRef = (ref: AbortController) => {
+    abortController = ref;
+  };
+
+  const getIsStreamInProgress = () => {
+    if (!abortController || abortController?.signal.aborted) {
+      abortController = undefined;
+      return false;
+    }
+    return true;
+  };
+
+  const subscribeToSearchRequests = () => {
+    const strictListener = () => {
+      const state = getState();
+      const requestId = state.search.requestId;
+      const streamId =
+        state.search.extendedResults.generativeQuestionAnsweringId;
+
+      if (lastRequestId !== requestId) {
+        lastRequestId = requestId;
+        abortController?.abort();
+        dispatch(resetAnswer());
+      }
+
+      const isStreamInProgress = getIsStreamInProgress();
+      if (!isStreamInProgress && streamId && streamId !== lastStreamId) {
+        lastStreamId = streamId;
+        dispatch(
+          streamAnswer({
+            setAbortControllerRef,
+          })
+        );
+      }
+    };
+    return engine.subscribe(strictListener);
+  };
+
+  subscribeToSearchRequests();
+
+  return {
+    ...controller,
+
+    get state() {
+      return getState().generatedAnswer;
+    },
+
+    retry() {
+      dispatch(executeSearch(logRetryGeneratedAnswer()));
```

Comment:
Are we retriggering the entire search for retry? If yes would it be possible to distinguish this in Analytics and guessing it is not possible to just retry gen-qa without retriggering the search?

Reply:
That's exactly what the `logRetryGeneratedAnswer` function does. The analytics event logged with this retry will specify an `actionCause` indicating that this is a retry for gen-qa.
Unfortunately it is not possible to try regenerating a stream without re-executing a query to the search API like this.